### ISSUE TITLE
Functions created in scripts are not immediately available after the script

### ISSUE THREAD

**Issue by patrickf3139** [fish-shell/fish-shell#5176]
Fish version: 2.7.1
Operating system: Mac
$TERM: xterm-256color
Tested without external plugins: Yes

Repro steps:
1. Create a file called `repro.fish` with the following content
```
echo "function proof; echo 'yes it works' ; end" > ~/.config/fish/functions/proof.fish
```
2. execute the file
```
fish repro.fish
```
3. try to execute the command just created, `proof`, and notice it fails
````
proof
fish: Unknown command 'proof'
```
4. open up a new fish shell and notice now `proof` works
```
fish
proof
yes it works
```

**Comment by krader1961** [comment]
Fish doesn't use inotify or whatever the platform provides for monitoring changes to the filesystem. For auto loaded functions it checks the directory(s) periodically. Something on the order of 10 to 15 seconds. So if you just wait a few seconds you'll find the new auto load script is recognized.

**Comment by patrickf3139** [comment]
Thanks you so much for answering Kurtis! Really appreciate this perplexing behavior being demystified. I literally wasted hours on this.

Anyway, in that case, I want to make a feature request for using the inotify system. Should I open a new ticket for that or can we just use this one?

**Comment by krader1961** [comment]
Note that fish runs on a large number of platforms. Many of which are not based on GNU Linux. Some of those provide mechanisms similar to Linux inotify. Some don't.  Too, implementing this even just on Linux will obviously complicate the code and it's not clear to me that is a good tradeoff. But I stopped contributing changes to the code a year ago so I'll leave it to the current project team to decide whether they want to turn this into an enhancement request.

**Comment by patrickf3139** [comment]
Thank you once again for your explanation. Yes, that would definitely complicate the codebase without a good layer of abstraction over the OS, which must be quite hard. In the meantime, time to hunt down the exact value of the refresh interval so I can build it into all my scripts...would appreciate anyone who knows it to point it out to me.

**Comment by faho** [comment]
>time to hunt down the exact value of the refresh interval so I can build it into all my script

Wait... what exactly are you doing that this matters?

The value is 15 seconds (src/autoload.cpp has "static const int kAutoloadStalenessInterval = 15;"), but that shouldn't matter, usually.

**Comment by patrickf3139** [comment]
My use case is really not that important. But all I'm doing is trying to automate my dotfiles, in particular my installation of the fisherman framework https://github.com/fisherman/fisherman/issues/439

**Comment by ridiculousfish** [comment]
When we try to autoload a function and it's not found, fish caches that fact. This cache expires after `kAutoloadStalenessInterval` seconds (currently 15).

**Comment by krader1961** [comment]
The "source route" means to execute `source ~/.config/fish/functions/proof.fish` to immediately execute the contents of the autoload function script. Thus bypassing the 15 second, worst case, delay before the new autoload script is recognized.

I say "worst case" because this is a situation where the average case is half the worst case. That is, 7.5 seconds rather than 15 seconds.

The problem with your attempt at automating installing *fisherman* plugins is that your sleep occurs before the first time you run `fisherman`. It needs to happen after it.

**Comment by faho** [comment]
>When we try to autoload a function and it's not found, fish caches that fact

@ridiculousfish: I guessed as much. The question is whether it should.

The problem here is this pattern:

```fish
if not type -q $something
    install $something
    $something
end
```

is basically guaranteed to fail because of that cache. And I don't really see where caching misses helps much. IMHO, autoloading should be as transparent as possible, and this isn't.

@krader1961: Adding the `sleep` after isn't great because that means you still have the first failure, which you probably don't want to have in a script.

**Comment by patrickf3139** [comment]
I don't think my use-case here is that important for automating my dotfiles, but I can see this being important for people in DevOps who want to standardize the tools available on all the machines. A script to do so would probably involve 
```fish
if not type -q $something
    install $something
    $something
end
```
If fish can't handle that, those people would have to rule out fish.

**Comment by mqudsi** [comment]
We can probably just predicate the cache on the function level, either each level in resets the view (for itself, not externally) or just the first level in (so when moving from interactive to sourced mode).

**Comment by faho** [comment]
>We can probably just predicate the cache on the function level

Why, though?

I'd still like to know why caching the fact that something doesn't exist helps? You're not expected to miss - that starts the command-not-found-handler every time, so the function lookup seems negligible. Even `type -q` isn't something you'd run multiple times in a row on the same function.

And to me it seems like that would solve the problem entirely, and make autoloading transparent.

**Comment by ridiculousfish** [comment]
> I'd still like to know why caching the fact that something doesn't exist helps? 

The problem is that functions can shadow builtins. If we didn't cache misses, we'd have to do multiple filesystem probes for every `set`, `echo`, etc. This would indeed be expensive.

However I think it makes perfect sense to expire this cache before `command_not_found`. That is, if a command doesn't resolve to a builtin, function, or external command, then try autoloading it again ignoring the cache.

**Comment by faho** [comment]
>The problem is that functions can shadow builtins

Ouch. I did not think of that! Yes, that makes complete sense.

>That is, if a command doesn't resolve to a builtin, function, or external command, then try autoloading it again ignoring the cache.

Doesn't that mean that new builtin wrappers won't be detected if the builtin has been used before? That's a much smaller failure case, but also more inconsistent. Maybe it's worth it?

**Comment by ridiculousfish** [comment]
Yes, if you shadow a builtin via an external process (nested fish instance, vim, etc) then we won't pick it up immediately, only after the cache expires. If you shadow a builtin in a way visible to fish including `funced`, `function`, etc we'll see it immediately. I think this is quite livable.

**Comment by kuncevic** [comment]
Just created my first plugin and notice that issue, seems a long lasting one. No solution so far?
